frontend/javascripts/dashboard/dataset/queries.tsx

```diff
@@ -0,0 +1,435 @@
+    };
+  }, [folderId]);
+
+  return queryData;
+}
+
+export function useCreateFolderMutation() {
+  const queryClient = useQueryClient();
+  const mutationKey = ["folders"];
+
+  return useMutation(([parentId, name]: [string, string]) => createFolder(parentId, name), {
+    mutationKey,
+    onSuccess: (newFolder) => {
+      queryClient.setQueryData(mutationKey, (oldItems: Folder[] | undefined) =>
+        (oldItems || []).concat([newFolder]),
+      );
+    },
+    onError: (err: any) => {
+      handleGenericError(err, "Could not create folder. Check the console for details");
+    },
+  });
+}
+
+export function useDeleteFolderMutation() {
+  const queryClient = useQueryClient();
+  const mutationKey = ["folders"];
+
+  return useMutation((id: string) => deleteFolder(id), {
+    mutationKey,
+    onSuccess: (deletedId) => {
+      queryClient.setQueryData(mutationKey, (oldItems: Folder[] | undefined) =>
+        (oldItems || []).filter((folder: Folder) => folder.id !== deletedId),
+      );
+    },
+    onError: (err: any) => {
+      handleGenericError(err, "Could not delete folder. Check the console for details");
+    },
+  });
+}
+
+export function useUpdateFolderMutation() {
+  const queryClient = useQueryClient();
+  const mutationKey = ["folders"];
+
+  return useMutation((folder: FolderUpdater) => updateFolder(folder), {
+    mutationKey,
+    onSuccess: (updatedFolder) => {
+      queryClient.setQueryData(mutationKey, (oldItems: FlatFolderTreeItem[] | undefined) =>
+        (oldItems || []).map((oldFolder: FlatFolderTreeItem) =>
+          oldFolder.id === updatedFolder.id
+            ? {
+                ...updatedFolder,
+                parent: oldFolder.parent,
+              }
+            : oldFolder,
+        ),
+      );
+      queryClient.setQueryData(["folders", updatedFolder.id], updatedFolder);
+    },
+    onError: (err: any) => {
+      handleGenericError(err, "Could not update folder. Check the console for details");
+    },
+  });
+}
+
+export function useMoveFolderMutation() {
+  const queryClient = useQueryClient();
+  const mutationKey = ["folders"];
+
+  return useMutation(
+    ([folderId, newParentId]: [string, string]) => moveFolder(folderId, newParentId),
+    {
+      mutationKey,
+      onMutate: ([folderId, newParentId]) => {
+        // Optimistically update the folder with the new parent.
+        const previousFolders = queryClient.getQueryData(mutationKey);
+        queryClient.setQueryData(mutationKey, (oldItems: FlatFolderTreeItem[] | undefined) =>
+          (oldItems || []).map((oldFolder: FlatFolderTreeItem) =>
+            oldFolder.id === folderId
+              ? {
+                  ...oldFolder,
+                  parent: newParentId,
+                }
+              : oldFolder,
+          ),
+        );
+        return {
+          previousFolders,
+        };
+      },
+      onSuccess: (updatedFolder, [folderId, newParentId]) => {
```

Comment:
It's not strictly necessary, but I thought, if there is a fresh folder instance from the server, why not use it, as it could have been changed by someone else (e.g., renamed). However, I DRYed the code a bit to avoid the redundancy :)